src/renderers/shared/fiber/ReactFiberCommitWork.js

```diff
@@ -462,7 +463,7 @@ module.exports = function<T, P, I, TI, C, CX>(
     }
     const ref = finishedWork.ref;
     if (ref) {
-      const instance = finishedWork.stateNode;
+      const instance = getPublicInstance(finishedWork.stateNode);
```

Comment:
`finishedWork.stateNode` on a a class component will be the instance of the class. That doesn't match the generic types `I | TI`. I think it only accidentally works on runtime now because you assume they won't have a `.tag` field with the string `INSTANCE` so they fallback.

You can make separate branches by switching on `finishedWork.tag` and only use `getPublicInstance` on the branch on `HostComponent`.